flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionManager.java

```diff
@@ -122,14 +123,9 @@ void onConsumedPartition(ResultPartition partition) {
 		}
 	}
 
-	public boolean areAllPartitionsReleased() {
+	public Collection<ResultPartitionID> getUnreleasedPartitions() {
 		synchronized (registeredPartitions) {
-			for (ResultPartition partition : registeredPartitions.values()) {
-				if (!partition.isReleased()) {
-					return false;
-				}
-			}
-			return true;
+			return registeredPartitions.keySet();
```

Comment:
The behavior here is different from before. We should confirm the criteria whether based on collection size or real partition state. It exists the scenario that the partition is already removed from the collection but the state is not released yet during`ResultPartitionManager#onConsumedPartition`. That means if we release the task executor immediately, it might lose chance for releasing the partition resource gracefully. WDYT?

Reply:
Ok, the behaviour did not actually change but this problem was already before :) If we remove partition in `ResultPartitionManager#onConsumedPartition` and concurrently continue to release it, the old `areAllPartitionsReleased` will still skip its state check because it is simply not  in the `registeredPartitions` anymore.

I think we have to fix it here by moving release under the lock as it is already done in `ResultPartitionManager.releasePartition`.

In general, I think `registeredPartitions` should reflect the release state of them. At the end, memory occupied by partition in `registeredPartitions` is also a resource to release so being in `registeredPartitions` and unreleased should be always the case. If partition is released/unregistered, there is no point to keep in `registeredPartitions` either.

Reply:
Yes, the `partition#release` should also put under the synchronized in `ResultPartitionManager#onConsumedPartition`, then it could solve this issue.